Restrict requests based on origin

ericmj opened the pull request:
No docs yet, sending PR to discuss implementation.

ericmj commented on mix.lock:
```diff
@@ -10,4 +10,4 @@
   "ranch": {:hex, :ranch, "1.0.0"},
   "redo": {:git, "git://github.com/heroku/redo.git", "b25e7bd3b197564192069d609d1e8e02183b71a6", []},
-  "websocket_client": {:git, "git://github.com/jeremyong/websocket_client.git", "2b8d9805306d36f22330f432ae6472f1f2625c30", []}}
+  "websocket_client": {:git, "git://github.com/jeremyong/websocket_client.git", "48c118682292f2e4d80491161134a665525c4934", []}}
```
Updating dependency to stop it from sending "ws://127.0.0.1" origin header.

josevalim commented on lib/phoenix/channel/transport.ex:
```diff
@@ -206,4 +206,34 @@ defmodule Phoenix.Channel.Transport do
     :ok
   end
+
+  def origin_allowed?(nil, _) do
```
We should document this. The idea is that: if the client does not send an origin, we are not going to check for it. However, this is not a problem with browsers because they all send the origin, correct?

ericmj replied:
Not sure if they send origin if it's not a cross-origin request.

josevalim replied:
That too. They don't if it is not cross-origin.

josevalim commented on lib/phoenix/channel/transport.ex:
```diff
@@ -206,4 +206,34 @@ defmodule Phoenix.Channel.Transport do
+                 compare?(origin.port, allowed.port)
+
+      # "example.com" parses into path so compare it instead of host
+      if allowed.host == nil do
```
Let's ask the user to provide proper URLs so we don't need this check.

ericmj replied:
You're right, it's not needed if using "//foo.bar" urls.

josevalim replied:
We should even raise if there is no host.

ericmj commented:
* Implemented discussed changes in comments and IRC.
* Moved origin check so code can be shared by all transports.
* Documented LongPoller and :transports Endpoint configuration.

Looks good?

josevalim commented on lib/phoenix/channel/transport.ex:
```diff
@@ -206,4 +206,40 @@ defmodule Phoenix.Channel.Transport do
     :ok
   end
+
+  @doc """
```
Doc missing. :) Also, can we unit test this function directly instead of testing all combinations through websockets and long polling? The unit tests will be much, much faster.

ericmj replied:
Fixed and added unit test.

chrismccord commented:
Thanks @ericmj! Will your xdomain PR require any changes based on this?

ericmj commented:
@chrismccord already rebased :)
